20230308/mike.java

```diff
@@ -0,0 +1,60 @@
+/*
+Use Binary Search to solve it, just same as recent daily questions.
+left can be 0 or 1, right could be max(piles) because 1 hour can only eat 1 stack.
+
+Edge case need to be careful:
+1. piles's length = 1 or 2
+*/
+class Solution {
+    int[] mPiles;
+    int mH;
+
+    public int minEatingSpeed(int[] piles, int h) {
+        Arrays.sort(piles);
+        mPiles = piles;
+        mH = h;
+        int left = 0;
+        int right = piles[piles.length-1];
+        if(left == right) {
+            int sum = 0;
+            for(int a: piles) {
+                sum+=a;
+            }
+
+            if(sum % h == 0) {
+                return sum/h;
+            } else {
+                return (sum/h)+1;
+            }
+        }
```

Comment:
if i understand correctly, `left` and `right` represent the minimum and maximum speed, respectively. then, `0` should be an invalid number.

furthermore, according to the constraints of the problem, `max` won't be 0 at any time.

> 1 <= piles[i] <= 109

as a result, `if (0 == 1)` will not be evaluated to true. in other words, this block of code is redundant.

Reply:
Suggest you to watch this video:

https://youtu.be/JuDAqNyTG4g

It offers an template to solve Binary Search problem. And there is the derivation in the video.
It use "partition" to describe BS and I think it's really useful.
But to let while(left +1 == right) be true, we should make our left and right starting from outside.

The video helps me many times~